File: javascript/easy/2692-take-gifts-from-the-richest-pile.js
/**
 * @param {number[]} gifts
 * @param {number} k
 * @return {number}
 */
var pickGifts = function (gifts, k) {
  const maxHeap = gifts.map(gift => -gift);
  maxHeap.sort((a, b) => a - b);

  for (const _ of Array(k)) {
    const maxGift = -maxHeap.shift();
    const newGift = Math.floor(Math.sqrt(maxGift));
    maxHeap.push(-newGift);
    maxHeap.sort((a, b) => a - b);
  }

  return maxHeap.reduce((sum, value) => sum - value, 0);
};

File: python/easy/2692-take-gifts-from-the-richest-pile.py
from typing import List
from heapq import heapify, heappop, heappush
from math import floor


class Solution:
    def pickGifts(self, gifts: List[int], k: int) -> int:
        h = [-x for x in gifts]
        heapify(h)
        for _ in range(k):
            x = -heappop(h)
            heappush(h, -int(floor(x**0.5)))
        return abs(sum(h))
